src/components/DatePickerRange.react.js

```diff
@@ -71,6 +71,10 @@ export default class DatePickerRange extends Component {
                 });
             }
         }
+
+        if (clearable && !start_date && !end_date) {
+            setProps({start_date: null, end_date: null});
+        }
```

Comment:
This fix correctly handles the `x` action case - only additional fix is that it might setting one of the date props superfluously if already / still `None`.

Leaving the scope of this specific issue / fix and looking into the logic of this component as a whole. Multiple scenarios seem wrongly implemented and I think we should use this opportunity to address them instead of just fixing this one case.

For example, setting `end_date` first with `updatemode=bothdates` will trigger setProps, setting `start_date` afterwards will not, even though both values are now set -- there's an assumption here that start is always set before end...; manually resetting a value with `clearable=True` is possible if only one value is set but not if both are, etc.

@alexcjohnson My feeling is we should pay the price and fix these components in depth as we work on them. I'm almost positive many other comps will exhibit deep logic flaws as we start going through them. Would rather fix them for real than make 'cosmetic' fixes that addresses whatever specific issue was logged.

[Warning: Wall of text...]

Four props are involved in the update logic: `start_date`, `end_date`, `clearable`, `updatemode`.

The possible operation / cases the user might trigger are:
With `clearable=True|False`
- set start_date w/ end_date unset
- set end_date w/ start_date unset
- set start_date w/ end_date set
- set end_date w/ start_date set

With `clearable=True`
- manually unset start_date w/ end_date unset
- manually unset end_date w/ start_date unset
- manually unset start_date w/ end_date set
- manually unset end_date w/ start_date set
- clear w/ start_date set, end_date unset
- clear w/ start_date unset, end_date set
- clear w/ start_date set, end_date set

All combinations and expected behaviors, hopefully without mistakes:

`clearable=False` + `updatemode=singledate`
- set start_date w/ end_date unset -> setProps called w/ start_date only
- set end_date w/ start_date unset -> setProps called w/ end_date only
- set start_date w/ end_date set -> setProps called w/ start_date only
- set end_date w/ start_date set -> setProps called w/ end_date only
- manually unset start_date w/ end_date unset -> does not unset
- manually unset end_date w/ start_date unset -> does not unset
- manually unset start_date w/ end_date set -> does not unset
- manually unset end_date w/ start_date set -> does not unset
- clear w/ start_date set, end_date unset -> not available
- clear w/ start_date unset, end_date set -> not available
- clear w/ start_date set, end_date set -> not available

`clearable=False` + `updatemode=bothdates`
- set start_date w/ end_date unset -> state updated, no setProps call
- set end_date w/ start_date unset -> state updated, no setProps call
- set start_date w/ end_date set -> setProps called w/ start_date and end_date
- set end_date w/ start_date set -> setProps called w/ start_date and end_date
- manually unset start_date w/ end_date unset -> does not unset
- manually unset end_date w/ start_date unset -> does not unset
- manually unset start_date w/ end_date set -> does not unset
- manually unset end_date w/ start_date set -> does not unset
- clear w/ start_date set, end_date unset -> not available
- clear w/ start_date unset, end_date set -> not available
- clear w/ start_date set, end_date set -> not available

`clearable=True` + `updatemode=singledate`
- set start_date w/ end_date unset -> setProps called w/ start_date only
- set end_date w/ start_date unset -> setProps called w/ end_date only
- set start_date w/ end_date set -> setProps called w/ start_date only
- set end_date w/ start_date set -> setProps called w/ end_date only
- manually unset start_date w/ end_date unset -> unsets, setProps called with start_date only
- manually unset end_date w/ start_date unset -> unsets, setProps called with end_date only
- manually unset start_date w/ end_date set -> unsets, setProps called with start_date only
- manually unset end_date w/ start_date set -> unsets, setProps called with end_date only
- clear w/ start_date set, end_date unset -> unsets, setProps called with start_date
- clear w/ start_date unset, end_date set -> unsets, setProps called with end_date
- clear w/ start_date set, end_date set -> unsets, setProps called with start_date and end_date

`clearable=True` + `updatemode=bothdates`
- set start_date w/ end_date unset -> state updated, no setProps call
- set end_date w/ start_date unset -> state updated, no setProps call
- set start_date w/ end_date set -> setProps called w/ start_date and end_date
- set end_date w/ start_date set -> setProps called w/ start_date and end_date
- manually unset start_date w/ end_date unset -> unsets, setProps called with start_date
- manually unset end_date w/ start_date unset -> unsets, setProps called with end_date
- manually unset start_date w/ end_date set -> unsets, no setProps call
- manually unset end_date w/ start_date set -> unsets, no setProps call
- clear w/ start_date set, end_date unset -> unsets, setProps called with start_date
- clear w/ start_date unset, end_date set -> unsets, setProps called with end_date
- clear w/ start_date set, end_date set -> unsets, setProps called with end_date and start_date

NB. The above considers that `bothdates` triggers setProps call if either both values are defined or both vaules are None.

Reply:
I OK to increase the scope, but prefer to close this PR and start a `wall` issue

Reply:
Alright. In that case just the first part of the comment, before the divider, applies.

Reply:
Please open a `wall` issue for the follow up :)